include/xtensor/xfft.hpp

```diff
@@ -0,0 +1,317 @@
+                const std::size_t n = data.size();
+                std::size_t m = 1;
+                while (m / 2 <= n) 
+                {
+                    m *= 2;
+                }
+
+                // Trignometric table
+                auto expTable = xt::xtensor<std::complex<T>, 1>::from_shape({ n });
+                xt::xtensor<std::size_t, 1> i = xt::pow(xt::linspace<std::size_t>(0, n - 1, n), 2);
+                i %= (n * 2);
+                
+                //I would like to do this but there appears to be a compiler error in CPP14
+                //Works in CPP17 and MSVC CPP14
+                //Linker error when using numeric_constants in this header
+                //auto angles = xt::eval(::xt::numeric_constants<T>::PI * i / n);
+                auto angles = xt::eval(pi<T>() * i / n);
+                auto j = std::complex<T>(0, 1);
+                expTable = xt::exp(-angles * j);
+
+                // Temporary vectors and preprocessing
+                xt::xarray<std::complex<T>> av = xt::empty<std::complex<T>>({ m });
+                xt::view(av, xt::range(0, n)) = data * expTable;
+
+
+                xt::xarray<std::complex<T>> bv = xt::empty<std::complex<T>>({ m });
+                xt::view(bv, xt::range(0, n)) = xt::conj(expTable);
+                xt::view(bv, xt::range(-n + 1, xt::placeholders::_)) = xt::view(xt::conj(xt::flip(expTable)), xt::range(xt::placeholders::_, - 1));
+
+                // Convolution
+                auto cv = xt::fft::fft_convolve<T>(av, bv);
+
+                return xt::eval(xt::view(cv, xt::range(0, n)) * expTable);
+            }
+
+            template<
+                typename T
+            >
+            auto fft_impl(xt::xarray<std::complex<T>> data)
+            {
+                //loop through all the next highest axis
+
+                // Is power of 2
+                if (detail::is_power2(data))
+                {
+                    return detail::transform_radix2<T>(std::move(data));
+                }
+                //More complicated algorithm for arbitrary sizes
+                else 
+                {
+                    return detail::transform_bluestein<T>(std::move(data));
+                }
+            }
+
+            template<
+                typename T, 
+                typename E1
+            >
+            auto fft(E1&& data, std::ptrdiff_t axis = -1)
+            {
+                //check the length of the data on that axis
+                const std::size_t n(data.shape(axis));
+                if (n == 0) {
+                    XTENSOR_THROW(std::runtime_error, "Cannot take the FFT along an empty dimention");
+                }
+
+                //select the axis
+                //create the return type
+                //this can be made smarter to use floats or double for small speed up
+                xt::xarray<std::complex<T>> complex_out = xt::zeros<std::complex<T>>(data.shape());
```

Comment:
It would be nice to keep an xtensor if you can. I actually don't know what is the proper constructor to choose automatically. @JohanMabille could you comment?

Reply:
That would not work: a typecast `T` -> `std::complex<T>` is involved

Reply:
Ah indeed. I forgot the cast: 

```cpp
auto complex_out = xt::zeros_like(xt::cast<std::complex<T>>(data));
```

Reply:
Just out of curiosity: will the case result in a copy that is immediately disposed of?

Reply:
The cast is lazy, it returns an xfunction that applies the cast to scalars on demand.